src/Symfony/Component/HttpClient/Response/CurlResponse.php

```diff
@@ -0,0 +1,268 @@
+        return $this;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getInfo(string $type = null)
+    {
+        if (!$info = $this->finalInfo) {
+            self::perform($this->multi);
+            $info = array_merge($this->info, curl_getinfo($this->handle));
+            $info['url'] = $this->info['url'] ?? $info['url'];
+            $info['redirect_url'] = '' === $info['redirect_url'] ? null : $info['redirect_url'];
+
+            // workaround curl not subtracting the time offset for pushed responses
+            if (isset($this->info['url']) && $info['start_time'] / 1000 < $info['total_time']) {
+                $info['total_time'] -= $info['starttransfer_time'] ?: $info['total_time'];
+                $info['starttransfer_time'] = 0.0;
+            }
+
+            if (!\in_array(curl_getinfo($this->handle, CURLINFO_PRIVATE), ['headers', 'content'], true)) {
+                $this->finalInfo = $info;
+            }
+        }
+
+        return null !== $type ? $info[$type] ?? null : $info;
+    }
+
+    public function __destruct()
+    {
+        try {
+            if (null === $this->timeout || isset($this->info['url'])) {
+                return; // pushed response
+            }
+
+            if ('content' === $waitFor = curl_getinfo($this->handle, CURLINFO_PRIVATE)) {
+                $this->close();
+            } elseif ('headers' === $waitFor) {
+                curl_setopt($this->handle, CURLOPT_PRIVATE, 'destruct');
+            }
+
+            $this->doDestruct();
+        } finally {
+            $this->close();
+
+            // Clear local caches when the only remaining handles are about pushed responses
+            if (\count($this->multi->openHandles) === \count($this->multi->pushedResponses)) {
+                $this->multi->pushedResponses = [];
+                // Schedule DNS cache eviction for the next request
+                $this->multi->dnsCache[2] = $this->multi->dnsCache[2] ?: $this->multi->dnsCache[1];
+                $this->multi->dnsCache[1] = $this->multi->dnsCache[0] = [];
+            }
+        }
+    }
+
+    private function close(): void
+    {
+        unset($this->multi->openHandles[$this->id], $this->multi->handlesActivity[$this->id]);
+        curl_multi_remove_handle($this->multi->handle, $this->handle);
+        curl_setopt_array($this->handle, [
+            CURLOPT_PRIVATE => '',
+            CURLOPT_NOPROGRESS => true,
+            CURLOPT_PROGRESSFUNCTION => null,
+            CURLOPT_HEADERFUNCTION => null,
+            CURLOPT_WRITEFUNCTION => null,
+            CURLOPT_READFUNCTION => null,
+            CURLOPT_INFILE => null,
+        ]);
+    }
+
+    private static function schedule(self $response, array &$runningResponses): void
+    {
+        if ('' === curl_getinfo($ch = $response->handle, CURLINFO_PRIVATE)) {
+            // no-op - response already completed
```

Comment:
This is an odd if, could be changed into a guard clause with early return

Reply:
Hello, 
You should review this PR if it's about the component itself. ;)
https://github.com/symfony/symfony/pull/30413

Reply:
I can't tell if your changes are a PR to the same branch including the other changes. At this point it's impossible to properly review until that one is merged.

@nicolas-grekas, as you're the author of that PR, this comment and the one below are probably for you :smile: